### ISSUE TITLE
Overweight on_runtime_upgrade could be problematic

### ISSUE THREAD

**Issue by ferrell-code** [paritytech/substrate#10064]
https://github.com/paritytech/substrate/blob/72a8f9fc6e25a82636b7d4b35ce98f79ec29ec38/frame/tips/src/migrations/v4.rs#L80

In the recent storage migrations of pallets (example above) we return the `MAXIMUM_BLOCK_WEIGHT` for the `migrate()` method. This creates a weird output when you migrate multiple instances in the runtime, example `try-runtime` output:
```
2021-10-18 19:34:55 TryRuntime_on_runtime_upgrade executed without errors. Consumed weight = 5500100000000, total weight = 500000000000 (11.0002)
```

My understanding is that in this case weight being greater than max-weight doesn't matter as it calls `register_extra_weight_unchecked` for `on_runtime_upgrade` (executes regardless of weight) and then that means nothing else will be executed that block. 

But doesn't that mean that if a storage migration happened to take longer than a block to execute using `on_runtime_upgrade()` that would stop block production as it's weight is unchecked? I could be confused so any clarification would be helpful :)

**Comment by bkchr** [comment]
> that would stop block production as it's weight is unchecked?

I don't get this question.

**Comment by bkchr** [comment]
> This is not true for parachain context right? 

Yes that is right, but this wasn't asked ;) I mean if we speak about parachains, then it would be nice to directly mentioning this.

> So we still need to know the real weight, which the migrations are not providing.

Ahh so you want to use the weight outputted by try-runtime to check that you are in the limits?

**Comment by ferrell-code** [comment]
> Yes that is right, but this wasn't asked ;) I mean if we speak about parachains, then it would be nice to directly mentioning this.

Yeah that was my bad, should of just led with the concrete example.

My understanding is that there is a 2 part solution

1. Get try-runtime to benchmark the storage migration and return actual weights
2. Create a mechanism for `on_runtime_upgrade()` to execute over the course of multiple blocks (this may be nice to have in the future even if in this particular case it is unnecessary)
